core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -377,7 +377,7 @@ private synchronized DocumentMapper merge(DocumentMapper mapper, MergeReason rea
         this.hasNested = hasNested;
         this.fullPathObjectMappers = fullPathObjectMappers;
         this.parentTypes = parentTypes;
-        this.allEnabled = mapper.allFieldMapper().enabled();
+        this.allEnabled |= mapper.allFieldMapper().enabled();
```

Comment:
can you add an assertion / comment here why this is safe to do? (i.e. we never go from enabled to disabled as the all_field enabledness of a type cannot be changed and types cannot be removed)